libcudacxx/include/cuda/std/detail/libcxx/include/barrier

```diff
@@ -384,9 +382,8 @@ private:
     static _LIBCUDACXX_INLINE_VISIBILITY constexpr
     uint64_t __init(ptrdiff_t __count) noexcept
     {
-#if (_LIBCUDACXX_DEBUG_LEVEL >= 2)
-        _LIBCUDACXX_DEBUG_ASSERT(__count >= 0);
-#endif
+        // We would like to assert that __count >= 0 in debug mode, but this
+        // fails to compile due to constexpr.
```

Comment:
What's the error?

@miscco @griwes the debug assert macro needs to support constant evaluation somehow in some mode, e.g., via some combination of `if constexpr`, throwing, etc.

Reply:
So technically all asserts should work in a constexpr context.

The `_LIBCUDACXX_DEBUG_ASSERT` internally has a condition `is_constant_evaluated() || COND` so it should always pass during constexpr evaluation. If it does not then it is a compiler issue and we should guard against those older compilers rather than not checking.

I would like to note that all non-constexpr functions should rather use `_LIBCUDACXX_ASSERT`

Reply:
> so it should always pass during constexpr evaluation

So if the assert should have failed, if it happens to run during constant evaluation then it wont fail ?

Reply:
Yeah, compiler support ...

Reply:
Why would there be a compiler bug against gcc-6? And why would that be fixed?

Reply:
The effects of assert is preventing the evaluation of the next program step on failure.
An implementation that does not achieve this, is broken.

You seem to be claiming that there are some bugs on some compilers that prevent such an implementation from ever existing, giving gcc-6 as an example. Weird, cause the following works just fine:  https://cpp.godbolt.org/z/MxcrxxKG1

Repeating my question: can you back up the claim that compiler support prevents a correct implementation of assert that achieve its main goal? For example, do we have a comment on our assert implementation describing its compiler support and current workaround for known compiler bugs.

If gcc-6 has a bug in its constexpr implementation that makes the above unreliable, but gcc-7 does not have that bug, then there is probably a bug report in gcc bugzilla about it from back then, which may contain a workaround.

> Why would there be a compiler bug against gcc-6? And why would that be fixed?

I have no idea what these questions have to do with anything.

Reply:
Resolving this, so discussion can continue in issue https://github.com/NVIDIA/cccl/issues/422